src/core/manager.c

```diff
@@ -2789,11 +2800,33 @@ static int manager_dispatch_notify_fd(sd_event_source *source, int fd, uint32_t
                 }
         }
 
+        _cleanup_(pidref_done) PidRef pidref = PIDREF_NULL;
+
         if (!ucred || !pid_is_valid(ucred->pid)) {
                 log_warning("Received notify message without valid credentials. Ignoring.");
                 return 0;
         }
 
+        if (pidfd >= 0)
+                r = pidref_set_pidfd_consume(&pidref, TAKE_FD(pidfd));
+        else
+                r = pidref_set_pid(&pidref, ucred->pid);
+        if (r < 0) {
+                if (r == -ESRCH)
+                    log_debug_errno(r, "Notify sender died before message is processed. Ignoring.");
+                else
+                    log_warning_errno(r, "Failed to pin notify sender, ignoring message: %m");
+                return 0;
+        }
+
+        if (pidref.pid != ucred->pid) {
+                assert(pidref.fd >= 0);
```

Comment:
this also feels like it should be a check rather than an assert

Reply:
If this is ever reached, the PidRef must be initialized from pidfd, so this IMO should be an assertion?

Reply:
It could be a system where pidfds are not available or blocked

Reply:
Hmm? Could SCM_PIDFD work in that case? By "the PidRef must be initialized from pidfd" I mean that if we don't have SCM_PIDFD, we call `pidref_set_pid(ucred->pid)`, and `pidref.pid` always equals to `ucred->pid`. So, to reach this branch, the PidRef must come from SCM_PIDFD, which means pidfd is in use.

Reply:
ok, I'm just worried about seccomp/LSM blocking things in weird ways